pkg/controller/barrier.go

```diff
@@ -0,0 +1,223 @@
+	// condition to wake goroutines that may be waiting for a reservation slot.
+	release(*sync.Cond)
+}
+
+// TaskBarrier is in incremental barrier that makes it easy to create multiple
+// async operations and wait for them as a group. It allows creation to be rate
+// limitted by a Limiter. It reports errors to a ErrorPolicy, which aggregates
+// the errors and decides whether the barrier should continue or bail out
+// early. It takes care of error prone requirements, such as remembering to
+// count the start tasks with a waitgroup and synchronizing completion
+// reporting out of the task loop.
+type TaskBarrier struct {
+	// cond guards all state including the policy.
+	cond *sync.Cond
+	// Policy is the ErrorPolicy that the task barrier will apply to errors
+	// returned by tasks.
+	Policy ErrorPolicy
+	// Limiter is the limiter that the task barrier will use to limit the
+	// creation of tasks.
+	Limiter Limiter
+	// the barrier becomes done when:
+	// * a call to policy returns false, indicating that the barrier should not
+	//   continue creating tasks.
+	// * on the first return of Wait()
+	done bool
+	// each active task places a hold on the task barrier. Wait() will not return
+	// until all holds are removed.
+	holds int
+}
+
+// NewTaskBarrier creates a task barrier with a limiter that never blocks and
+// ignores all errors. The Policy and Limiter can be replace before the task
+// barrier is used.
+func NewTaskBarrier() *TaskBarrier {
+	return &TaskBarrier{
+		cond:    sync.NewCond(&sync.Mutex{}),
+		Policy:  ignoreErrorPolicy,
+		Limiter: NewNoLimiter(),
+	}
+}
+
+// Go requests that the barrier start a new task. It returns false if the
+// barrier is unable to start the task (i.e. the barrier is already done).  In
+// most circumstances, the return value can be safely ignored.
+func (tb *TaskBarrier) Go(task func() error) bool {
+	tb.cond.L.Lock()
+	defer tb.cond.L.Unlock()
+
+	// Wait for the limiter to give the goahead or the task barrier to
+	// transition to done.
+	for !tb.Limiter.reserve() && !tb.done {
+		tb.cond.Wait()
+	}
+
+	// if the task done, don't do anything.
+	if tb.done {
+		return false
+	}
+
+	// add a hold and start the task.
+	tb.holds++
+
+	go func() {
+		// Run the task, then under lock:
+		// * release the task's hold on the barrier
+		// * release the task's reservation back to the limiter
+		// * report the task's return to ErrorPolicy. If the ErrorPolicy indicates
+		//   that we shouldn't continue, set the barrier to done.
+		// * if holds drop to zero or the barrier is done, wake everyone up.
+		err := task()
+
+		tb.cond.L.Lock()
+		defer tb.cond.L.Unlock()
+
+		tb.holds--
+		tb.Limiter.release(tb.cond)
+
+		if ok := tb.Policy.Apply(err); !ok {
+			tb.done = true
+		}
+		if tb.holds == 0 || tb.done {
+			tb.cond.Broadcast()
+		}
+	}()
+	return true
+}
+
+// Wait waits for all holds tasks to complete.
+func (tb *TaskBarrier) Wait() {
```

Comment:
I think this is racy.
Imagine you have multiple tasks in flight, and more waiting on the limiter in func Go.
If all those in-flight tasks finish around the same time, Wait could unblock before any of the Go goroutines do and mark the barrier as done.
Then all those pending Go calls would bail.
And since most callers won't check the return value from Go, there can be tasks that get silently dropped.

Reply:
This is the intended behavior. The idea is that generally a user of this will call Go to prepare all tasks and then Wait in the same thread. Calls to Go after the barrier is complete have no effect. Suggestions on a better interface?

Reply:
Ack, point this out explicitly in a doc comment